vllm/v1/core/kv_cache_manager.py

```diff
@@ -36,73 +169,413 @@ def __init__(
         self.num_preallocate_tokens = num_preallocate_tokens
         self.num_preallocate_blocks = cdiv(num_preallocate_tokens, block_size)
 
-        self.free_block_ids = list(range(num_gpu_blocks))
-        self.req_to_block_ids: Dict[str, List[int]] = {}
-        self.ref_cnts = np.zeros(num_gpu_blocks, dtype=np.int32)
+        # A Block pool of all kv-cache blocks.
+        self.block_pool: List[KVCacheBlock] = [
+            KVCacheBlock(idx) for idx in range(num_gpu_blocks)
+        ]
```

Comment:
QQ: While it doesn't need to be in this PR, can we reserve block 0 for a special purpose (e.g., for padding)?

Just wondering if this change would be difficult after this PR gets merged.

Reply:
I don't think it won't be a problem, but could you elaborate a bit more so that I could help justify that?

Reply:
While it's not the case for V1 at the moment, when we have padded tokens, we have to handle them specially using a custom kernel (e.g., `reshape_and_cache_flash`). If we reserve one block for those tokens, we can just use PyTorch's `index_copy` without needing a special kernel. This trick is used by Gaudi in V0.

That being said, I think it's fine as long as it's not difficult to change this in the future.